calico-vpp-agent/connectivity/connectivity_server.go

```diff
@@ -352,3 +352,35 @@ func (s *ConnectivityServer) updateIPConnectivity(cn *common.NodeConnectivity, I
 		}
 	}
 }
+
+// ForceRescanState forces to rescan VPP state (ConnectivityProvider.RescanState()) for initialized
+// ConnectivityProvider of given type.
+// The usage is mainly for testing purposes.
+func (s *ConnectivityServer) ForceRescanState(providerType string) (err error) {
+	provider, found := s.providers[providerType]
+	if !found {
+		return fmt.Errorf("can't find connectivity provider of type %s", providerType)
+	}
+	provider.RescanState()
+	return nil
+}
+
+// ForceProviderEnableDisable force to enable/disable specific connectivity provider.
+// The usage is mainly for testing purposes.
+func (s *ConnectivityServer) ForceProviderEnableDisable(providerType string, enable bool) (err error) {
+	provider, found := s.providers[providerType]
+	if !found {
+		return fmt.Errorf("can't find connectivity provider of type %s", providerType)
+	}
+	provider.EnableDisable(enable)
+	return nil
+}
+
+// TODO get rid (if possible) of all this "Force" methods by refactor the test code
```

Comment:
I think it's ok for now to keep those around. It's not ideal, but I don't see an easy way to do introspection from the test package otherwise.

Reply:
> I don't see an easy way to do introspection from the test package otherwise.

As i wrote on the second line of that TODO, it could be done by just running the event loop and sending event messages. Basically that way is the code running in the wild. The 3 "force" methods can be removed that way:
1. execute provider.RescanState() by simple run of the event loop function (it does the rescanstate for all providers)
2. execute s.providers[WIREGUARD].EnableDisable(...) by sending proper common.FelixConfChanged event
3. modify nodeByAddr by sending common.PeerNodeStateChanged event

Do we need to run event loop in the background? Will it complicate tests? Does it need additional work? Yes, yes and yes. But we can get rid of the ugly "force" method.

Basically it is a tradeoff between nicer functionality implementation and easier test implementation.
So no "easy" way, but it is doable.